src/main/java/hudson/remoting/forward/CopyThread.java

```diff
@@ -16,10 +16,23 @@
     private final InputStream in;
     private final OutputStream out;
 
-    public CopyThread(String threadName, InputStream in, OutputStream out) {
+    public CopyThread(String threadName, InputStream in, OutputStream out, Runnable termination) {
+        this(threadName, in, out, termination, 0);
+    }
+
+    private CopyThread(String threadName, InputStream in, OutputStream out, Runnable termination, int previousTries) {
         super(threadName);
         this.in = in;
         this.out = out;
+        setUncaughtExceptionHandler((t, e) -> {
+            if (previousTries < 5) {
+                LOGGER.log(Level.FINE, "Uncaught exception in CopyThread " + t + ", retrying copy", e);
+                new CopyThread(threadName, in, out, termination, previousTries + 1).start();
```

Comment:
I played around here with trying to prevent data loss (or rewriting data) but I couldn't come up with a solid solution.  You can reset the `InputStream` but I could not figure out a way to determine how many bytes were successfully written to the output stream to restart the `InputStream` from there (short of literally writing one byte at a time).  I think that approach can still be used for errors on reading the `InputStream`, but I don't think that any reset will be safe if an exception is thrown while attempting to write.

All of this leads me to question if we should be trying to restart the thread from here if an exception is thrown while writing.  It may be better to do it from the `Runnable` passed into `CopyThread` so that the streams can be reinitialized.  I'm going to try to prototype that approach when I get back home tonight).